### DISCUSSION TITLE
Two questions about memory usage of virtual machines

### DISCUSSION THREAD

**Discussion by Verhoeckx** [cockpit-project/cockpit-machines#270] [Q&A]
I have two question about the memory usage of virtual machines:
1. Is the memory static of dynamic?
2. Is there a general how much memory you should assign to a virtual machine?

With static I mean the host system always uses all allocated memory (even when the virtual machine uses less).
With dynamic I mean the host systems only uses the memory that the virtual machine is actually using.

@KKoukiou: Do you have time to answer above questions :smiley: ?

**Comment by KKoukiou** [comment]
Hey @Verhoeckx,

1. Short answer is dynamic as we do use memory ballooning, read more about it for example here https://pmhahn.github.io/virtio-balloon/

2. Different OSes have different memory recommendations. [osinfo-db](https://gitlab.com/libosinfo/osinfo-db/) is the project that we fetch such information from, so that we can pick the choose the default values. In specific, if you create a VM through cockpit UI, after you fill the operating system field in the 'Create' dialog, the memory size and storage size values are prefilled with the minimum recommended values for the specific OS. 
If you try to insert any value less than the minimum you will see a warning in the UI as follows in the screenshot.
![Screen Shot 2021-07-30 at 15 20 27](https://user-images.githubusercontent.com/14921356/127659199-74ca8225-f193-4ee3-802c-b13c6f66bab2.png)

**Reply by Verhoeckx** [reply]
Thank you very much for the information :smiley: !
Exactly what I wanted to know!

So I have to interpret `Size` as the _maximum_ size the virtual machine can use.

Last question about memory:
What is the difference between `current` and `maximum` allocation in the `Nextcloud memory adjustment` dialog?
Do I have to interpret current as the _minimum_ allocation?

**Reply by KKoukiou** [reply]
For reference:
![Screen Shot 2021-07-30 at 16 02 34](https://user-images.githubusercontent.com/14921356/127664711-9e482bc2-b1dd-4dcb-b4e0-83862a97f164.png)


@Verhoeckx current memory, is the actual allocation of memory for the guest. This can be adjusted while the guest is running, so you can reduce the guest memory or increase it while the VM is running. 

Max memory is the maximum value that your current memory value can be increased to.

This needs to be documented inline in the dialog, I know it's not obvious at all.

More information here: https://libvirt.org/formatdomain.html#memory-allocation 
Current memory in the UI -> current memory in the docs
Max memory in the UI -> memory in these docs

**Reply by Verhoeckx** [reply]
Thanks!

If I understand it well, you could bring it back to one slider?

When the virtual machine is shut off: the memory can slide in either direction.
When the virtual machine is running: the memory can only be reduced.

Or maybe I'm oversimplifying it? And I'm also not sure if this programmatically possible.

_Anyway, thanks, I now understand how it works!_

Sorry for all the ideas :laughing: !

**Reply by KKoukiou** [reply]
This needs more thought for sure, I will transfer this discussion to an issue, to make sure to document this properly or simplify it. Thanks for the suggestion.
